reinitialize-pods/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+    let pod_evts = rt.watch_all::<Pod>(
+        watcher::Config::default()
+            .labels(DATA_PLANE_LABEL)
+            .fields(&format!("spec.nodeName={node_name}")),
+    );
+    let (tx, rx) = mpsc::channel(EVENT_CHANNEL_CAPACITY);
+    tokio::spawn(process_events(pod_evts, tx, metrics.clone()));
+
+    let client = rt.client();
+    tokio::spawn(process_pods(client, controller_pod_name, rx, metrics));
+}
+
+async fn process_events(
+    pod_evts: impl Stream<Item = watcher::Event<Pod>>,
+    tx: Sender<Pod>,
+    metrics: Metrics,
+) {
+    tokio::pin!(pod_evts);
+    while let Some(evt) = pod_evts.next().await {
+        tracing::trace!(?evt);
+        if let watcher::Event::Applied(pod) = evt {
+            let status = if let Some(ref status) = pod.status {
+                status.clone()
+            } else {
+                tracing::info!("Skipped, no status");
+                continue;
+            };
+
+            let terminated = status
+                .init_container_statuses
+                .unwrap_or_default()
+                .iter()
+                .find(|status| status.name == "linkerd-network-validator")
+                .and_then(|status| status.last_state.as_ref())
+                .and_then(|status| status.terminated.as_ref())
+                .map(|term| term.exit_code == UNSUCCESSFUL_EXIT_CODE)
+                .unwrap_or(false);
+
+            let evicted = status
+                .conditions
+                .as_ref()
+                .and_then(|conds| {
+                    conds.iter().find(|cond| {
+                        cond.reason
+                            .as_ref()
+                            .is_some_and(|reason| reason == CONDITION_EVICTED_REASON)
+                    })
+                })
+                .is_some();
+
+            if terminated && !evicted {
+                let namespace = pod.namespace().unwrap();
+                let name = pod.name_any();
+                // this avoids blocking the event loop
+                match tx.try_send(pod.clone()) {
+                    Ok(_) => {}
+                    Err(TrySendError::Full(_)) => {
+                        tracing::warn!(%namespace, %name, "Dropped event (channel full)");
+                        metrics.queue_overflow.inc();
+                    }
+                    Err(TrySendError::Closed(_)) => {
+                        tracing::warn!(%namespace, %name, "Dropped event (channel closed or dropped)")
+                    }
+                }
+            }
+        }
+    }
+}
+
+async fn process_pods(
+    client: Client,
+    controller_pod_name: String,
+    mut rx: Receiver<Pod>,
+    metrics: Metrics,
+) {
+    while let Some(pod) = rx.recv().await {
+        let namespace = pod.namespace().unwrap();
+        let name = pod.name_any();
+        let pods = kube::Api::<Pod>::namespaced(client.clone(), &namespace);
+        let evict_res = pods.evict(&name, &Default::default()).await;
```

Comment:
Eviction is a newer API, more graceful for terminating pods, as it respects PodDisruptionBudgets and terminationGracePeriodSeconds. This doesn't strictly apply to our case where the pod is already in a bad state to begin with, but it just feels to me like a more proper way compared to a straight resource deletion.

Reply:
It seems likely to me that Eviction could degrade the behavior of this controller -- won't pod disruption budgets prevent these pods from being deleted as quickly as possible in some cases? Why would a user want that behavior?

This isn't strictly a blocker, but I do think it would be good to be clearer about the functional trade-offs -- it's only a proper solution if the trade-offs are desirable :)

Reply:
Ok I'll revert back to simple deletion; there isn't really a use case here for anything other than killing these pods that haven't even had a chance to start properly :+1: